### ISSUE TITLE
584109C2 - Perfect Dark

### ISSUE THREAD

**Issue by KyouKeiKen** [xenia-project/game-compatibility#316]
Marketplace (http://marketplace.xbox.com/en-AU/Product/Perfect-Dark/66acd000-77fe-1000-9115-d802584109c2)

Tested on benvanik/xenia@b5469b5

Crashes on Product Identification screen

<h1>LOG

[LOG] (https://www.mediafire.com/?qlnrrxyz2r2qqhf)

<h1>Screenshots

![2015-12-13 12_32_22-xenia](https://cloud.githubusercontent.com/assets/6925024/11764879/2c6d7cca-a196-11e5-8a7a-596f29c9125a.jpg)

**Comment by sheppard66** [comment]
How did you launched it? I got only 1 file whitout extension and Xenia can't launch it.

**Comment by KyouKeiKen** [comment]
This one is a mix bag. I mean my friend can't get it to boot with his AMD card & Xenia would crash for me on Xbox Live Arcade Logo or Product Identification screen. I just did open file > (Bunch of letters & numbers) of Perfect Dark XBLA file

**Comment by sheppard66** [comment]
Ok thx for the answer, I have an AMD card so I understand why it doesn't work

**Comment by gibbed** [comment]
Now booting to gameplay as of benvanik/xenia@27c16b19362a11cd92de9f034477afd5156e577c in GL.

![perfectdark4](https://cloud.githubusercontent.com/assets/85170/16139967/43a576c2-3412-11e6-92ab-bc983fb71867.png)
![perfectdark5](https://cloud.githubusercontent.com/assets/85170/16139971/5541753e-3412-11e6-90ea-ce06084d6668.png)

**Comment by KyouKeiKen** [comment]
Here's some gameplay on benvanik/xenia@d586574. If u have AMD card, u most likely run into black screens at main menu.

https://www.youtube.com/watch?v=gULtY0fn_9Y

**Comment by ndo360** [comment]
On the recent version the game runs smoother but the video is distorted and most of it goes black.

**Comment by AmbientMalice** [comment]
Recent changes to Vulkan have dramatically improved the game. Now runs at full speed with minor movement glitches (Joanna seems to randomly sink to the floor before popping back up) and graphics equivalent to GL. https://youtu.be/5fuUVG3dj50

**Comment by ndo360** [comment]
That’s awesome! I wonder if it is possible to attempt multiplayer with some LAN emulator?

**Comment by hubslave** [comment]
You should change status to playable. Xenia from 3/4/18 plays it on my AMD radeon 7700 almost like on real HW (except when i select combat simulator,coop or counterop, it does not do anything, but i know what it could be. It’s most likely that dialog, which is used in all local multiplayer games to log in other controllers. Xenia outputs, the dialog was cancelled)
I can look at how xenia emulates those dialogs

**Comment by tjd2** [comment]
When i choose Combat Simulator in main menu, it shows loading... and then goes back to main menu. Can you please try to solve it?

**Comment by ORG4NX** [comment]
The game doesn't stop to back on pause when I'm doing a mission, can you take a look?

**Comment by illusion98** [comment]
Playable from start to finish but has noise on models.
 https://youtu.be/Zyc6F5wMcwQ

**Comment by Timecrash** [comment]
I've been following progress here for a while, but I'm still seeing a significant issue with an nvidia card. Apart from the noise on models, my framerate tanks the more of the game I play. It starts out fine, but frequently hits single digits if I restart a mission, or spend too much time in one.

I haven't really seen anyone report this, but it makes the game basically unplayable for me.

I've tried using Vulkan and D3D12, same issue.

If it helps, I've attached my config file:
[xenia.config.toml.zip](https://github.com/xenia-project/game-compatibility/files/4240529/xenia.config.toml.zip)

**Comment by needo2** [comment]
Not a bug but the game is unplayable with headless=true because first option in XBOX LIVE connect dialog does nothing

**Comment by Triang3l** [comment]
On Nvidia, vertex varying interpolation is inexact, for vertices with exactly -1 texcoords, some pixels get -0.999…, while the pixel shader expects exactly -1, so we're getting that noise on models.

**Comment by Timecrash** [comment]
Update for me: I don't know if it was a Xenia update or a config change I made, but I no longer get slowdown. Perfect Dark is now perfectly (ha) playable apart from the noise on models.

I've reattached my latest config; I made a couple changes to get DoDonPachi working.
[xenia.config.toml.zip](https://github.com/xenia-project/game-compatibility/files/5145601/xenia.config.toml.zip)

**Comment by Triang3l** [comment]
@Timecrash It's a hardware bug, or rather, imperfection (it's still apparently within the precision requirements of the D3D12 spec) in Nvidia hardware. I currently don't know how it can be fixed (apart from manually interpolating with SV_Barycentrics, but that requires a huge shader translator rework, and is not guaranteed to help).

**Comment by madmodder123** [comment]
> 
> 
> when i select combat simulator,coop or counterop, it does not do anything, but i know what it could be. It’s most likely that dialog, which is used in all local multiplayer games to log in other controllers. Xenia outputs, the dialog was cancelled.
> With older xenia (1.0.1219 texsplit i was getting major graphics bugs with glass, and invisible enemies)
> I can look at how xenia emulates those dialogs

Any updates to this issue? I can't get it working on the stable branch or the canary branch, after selecting "Play Game" and then, "Continue playing offline" I cant get into combat sim, co-cop, or counter-op. It just goes back to the current menu.

**Comment by madmodder123** [comment]
I managed to fix the issue loading the other game modes, a pull request has been submitted to resolve the issue.

@hubslave

**Comment by TheNewClassics** [comment]
Thanks madmodder for the fix. There's no way currently to get players 2-4 signed in when in those other modes, is there?

**Comment by katastic** [comment]
I've got a weird issue. If I load the game, I can could play with unlicensed demo, I played most of the first level.

But, after changing to license_mask = 1 or -1, now I go to the main menu and if I press "Play" it does nothing. The menu never leaves. It doesn't crash. It plays a menu sound. It just doesn't launch anything. 

![image](https://user-images.githubusercontent.com/11449593/164778065-e18fa99b-28fe-4500-ac35-e5eba6fe9858.png)

I'm going to look for cached files and delete them.

A second issue is, if I resize Xenia while the game is loading, Xenia crashes. So I have to wait for it to finish and get near/past the license [not the same 'license' as above] screen.

**Comment by katastic** [comment]
There appears to be another bug that affects scripting. Whenever there is a Joanna Dark text pop-up, it stays on forever. This usually occurs at the end of a level, or, just stays on and doesn't appear to affect level triggers so you just have a text popup all level. 

However, in the boat level (Pelagic) you're supposed to come upon 3 ship 'pilots' or whatever, Joanna says [see screenshot] then when she's done, you can go up to them and they react. One attempts to re-route the ship, a second will yell "traitor" and try to shoot him and you're supposed to disable/shoot him.

![image](https://user-images.githubusercontent.com/11449593/165205912-f0bbee2f-2e1a-4dbd-bf46-acc3ef703716.png)

**What happens in Xenia** is nothing. The voice plays, the text shows up, but never leaves, and so it never starts the NPC scripts. This is a required script. So the currently game is **unbeatable in Xenia**.

**Comment by theKraid** [comment]
Can confirm @katastic the Pelagic Mission is broken because of this

**Comment by abreiter** [comment]
please devs, i am thinkink about change my NVIDIA RTX 3070 to a AMD card just because of this.
is there any way to solve the graphic glitches in NVIDIA cards ?
thank you

**Comment by BowserN64** [comment]
> There appears to be another bug that affects scripting. Whenever there is a Joanna Dark text pop-up, it stays on forever. This usually occurs at the end of a level, or, just stays on and doesn't appear to affect level triggers so you just have a text popup all level.
> 
> However, in the boat level (Pelagic) you're supposed to come upon 3 ship 'pilots' or whatever, Joanna says [see screenshot] then when she's done, you can go up to them and they react. One attempts to re-route the ship, a second will yell "traitor" and try to shoot him and you're supposed to disable/shoot him.
> 
> ![image](https://user-images.githubusercontent.com/11449593/165205912-f0bbee2f-2e1a-4dbd-bf46-acc3ef703716.png)
> 
> **What happens in Xenia** is nothing. The voice plays, the text shows up, but never leaves, and so it never starts the NPC scripts. This is a required script. So the currently game is **unbeatable in Xenia**. Pressing A (use) on everything does nothing. Shooting near people does nothing. You can even knock out the middle and left guard without failing the mission. But it'll never proceed and finish the mission.
> 
> For reference here's two different guides explaining what's supposed to happen: ![image](https://user-images.githubusercontent.com/11449593/165206809-56379cbc-70d4-4f4a-8394-111823019a8c.png)
> 
> ![image](https://user-images.githubusercontent.com/11449593/165206916-d6a0e9c8-1399-40e7-8134-781873ed0b89.png)

I'm so glad I'm not the only one who was having this issue. You guys are awesome!

Fortunately, I have SOME GREAT NEWS! I think I may have found a workaround to this issue. I wasn't able to find any information about how to fix this, except the post from Katastic talking about the issue, so I went ahead and shot my own video tutorial/walkthrough. I have a condensed, more concise version and also a more detailed, longer version. In the longer version I played through the Pelagic II level in a couple different scenarios to test the reliability of the workaround. 

Here's the videos:

Condensed Version:
[https://youtu.be/N0b__S_B1YA](url)

Extended Version:
[https://youtu.be/nwtZUzhAFsA](url)

Here's the steps I took to work around the bug:

Workaround: 
NOTE: DO NOT COMPLETE OBJECTIVE #1(Disable primary power source) FIRST. 
THESE STEPS MUST BE DONE IN A CERTAIN ORDER.

1. Head to the Bridge.
2. Make guards shoot at you.
3. Johanna should say "Pull the plug on that, now!" and "Switch this thing off"
4. "Switch this thing off!" will stay on the screen and not disappear. Guard will not acknowledge her order. Do not panic.
5. Head back downstairs and turn off the power to the ship's reactor (Objective #1).
6. Objective #2 - Deactivate GPS and Autopilot will Complete.
7. Complete Objective 3 (Rendezvous with Elvis) and finish the level.

Tested on Xenia Canary 
Build: Canary_Experimental@d262214c1 on Oct 14, 2022

Played On Steam Deck

If I missed anything vital to getting this working for you guys, please let me know. Thanks for the amazing work on Xenia, and I hope this helps someone!

**Comment by BowserN64** [comment]
> There appears to be another bug that affects scripting. Whenever there is a Joanna Dark text pop-up, it stays on forever. This usually occurs at the end of a level, or, just stays on and doesn't appear to affect level triggers so you just have a text popup all level.
> 
> However, in the boat level (Pelagic) you're supposed to come upon 3 ship 'pilots' or whatever, Joanna says [see screenshot] then when she's done, you can go up to them and they react. One attempts to re-route the ship, a second will yell "traitor" and try to shoot him and you're supposed to disable/shoot him.
> 
> ![image](https://user-images.githubusercontent.com/11449593/165205912-f0bbee2f-2e1a-4dbd-bf46-acc3ef703716.png)
> 
> **What happens in Xenia** is nothing. The voice plays, the text shows up, but never leaves, and so it never starts the NPC scripts. This is a required script. So the currently game is **unbeatable in Xenia**. Pressing A (use) on everything does nothing. Shooting near people does nothing. You can even knock out the middle and left guard without failing the mission. But it'll never proceed and finish the mission.
> 
> For reference here's two different guides explaining what's supposed to happen: ![image](https://user-images.githubusercontent.com/11449593/165206809-56379cbc-70d4-4f4a-8394-111823019a8c.png)
> 
> ![image](https://user-images.githubusercontent.com/11449593/165206916-d6a0e9c8-1399-40e7-8134-781873ed0b89.png)

Well, I am happy to report, that, as of the 10/25/2022 build of Xenia_Canary, this seems to be NOW FIXED!!!! @katastic @theKraid 

The Pelagic II fix seems to be working on both AMD and Nvidia. Also, worthy of note, I was able to play through the whole game after I got through Pelagic II, on the previous build of Xenia, so we are SO CLOSE to PERFECTION on this game! 

Here's a video showing the bug fixed, playing through Pelagic II on Agent, on my Steam Deck on Windows 10! I also tested this on my Nvidia GTX 980M gaming laptop on Windows 11 with the same results: 

https://youtu.be/-kLYYn-7Xws

THIS IS SUPER EXCITING, because now it means we can all play the XBLA VERSION of Perfect Dark with REMASTERED music and GRAPHICS on our STEAM DECKS!

I wanna give a HUGE shout out to the developers of Xenia for fixing this bug, as I know it takes so many hours and countless amounts of hard work, so THANK YOU SO MUCH for all you do! I also wanna give a shout out to @foslerfer, as she watched the video I posted of the bug workaround, and from what I can gather possibly talked to the developers about this bug (but I can't be sure). It may be this was already in their list of bugs and issues to fix. Either way, THANK YOU SO MUCH! YOU GUYS ROCK MY WORLD!!!!

Now there's nothing left to do, but JAM OUT ON THIS AMAZING GAME!!!!

Uh-oh.

THERE'S SOMEONE HERE!!! :)

**Comment by striker3848** [comment]
> 

yeah in your video how did you fix the noise on the models? i am using dumped file 
named:
'35C1CDD22DD0D4E54B858859C0052124FFFAD17958'

**Comment by striker3848** [comment]
> > There appears to be another bug that affects scripting. Whenever there is a Joanna Dark text pop-up, it stays on forever. This usually occurs at the end of a level, or, just stays on and doesn't appear to affect level triggers so you just have a text popup all level.
> > However, in the boat level (Pelagic) you're supposed to come upon 3 ship 'pilots' or whatever, Joanna says [see screenshot] then when she's done, you can go up to them and they react. One attempts to re-route the ship, a second will yell "traitor" and try to shoot him and you're supposed to disable/shoot him.
> > ![image](https://user-images.githubusercontent.com/11449593/165205912-f0bbee2f-2e1a-4dbd-bf46-acc3ef703716.png)
> > **What happens in Xenia** is nothing. The voice plays, the text shows up, but never leaves, and so it never starts the NPC scripts. This is a required script. So the currently game is **unbeatable in Xenia**. Pressing A (use) on everything does nothing. Shooting near people does nothing. You can even knock out the middle and left guard without failing the mission. But it'll never proceed and finish the mission.
> > For reference here's two different guides explaining what's supposed to happen: ![image](https://user-images.githubusercontent.com/11449593/165206809-56379cbc-70d4-4f4a-8394-111823019a8c.png)
> > ![image](https://user-images.githubusercontent.com/11449593/165206916-d6a0e9c8-1399-40e7-8134-781873ed0b89.png)
> 
> Well, I am happy to report, that, as of the 10/25/2022 build of Xenia_Canary, this seems to be NOW FIXED!!!! @katastic @theKraid
> 
> 55877f4_canary_experimental https://github.com/xenia-canary/xenia-canary/releases/tag/experimental https://github.com/xenia-canary/xenia-canary/releases/tag/55877f4
> 
> The Pelagic II fix seems to be working on both AMD and Nvidia. Also, worthy of note, I was able to play through the whole game after I got through Pelagic II, on the previous build of Xenia, so we are SO CLOSE to PERFECTION on this game!
> 
> Here's a video showing the bug fixed, playing through Pelagic II on Agent, on my Steam Deck(An AMD-based computer) on Windows 10! I also tested this on my Nvidia GTX 980M gaming laptop on Windows 11 with the same results:
> 
> https://youtu.be/-kLYYn-7Xws
> 
> THIS IS SUPER EXCITING, because now it means we can all play the XBLA VERSION of Perfect Dark with REMASTERED music and GRAPHICS on our STEAM DECKS!
> 
> I wanna give a HUGE shout out to the developers of Xenia for fixing this bug, as I know it takes so many hours and countless amounts of hard work, so THANK YOU SO MUCH for all you do! I also wanna give a shout out to @foslerfer, as she watched the video I posted of the bug workaround, and from what I can gather possibly talked to the developers about this bug (but I can't be sure). It may be this was already in their list of bugs and issues to fix. Either way, THANK YOU SO MUCH! YOU GUYS ROCK MY WORLD!!!!
> 
> Now there's nothing left to do, but JAM OUT ON THIS AMAZING GAME!!!!
> 
> Uh-oh.
> 
> THERE'S SOMEONE HERE!!! :)

I tried using the build in your links but both still have noise on models

**Comment by BowserN64** [comment]
> > 
> 
> yeah in your video how did you fix the noise on the models? i am using dumped file named: '35C1CDD22DD0D4E54B858859C0052124FFFAD17958'

Hey! I didn't actually FIX the model noise. I was running the game on my Steam Deck, which uses AMD graphics, and apparently the model noise isn't a problem for AMD users. It's still a problem on Nvidia graphics though, which I tested with my Nvidia PC. In the video I think I mentioned I was using AMD (The Steam Deck uses a custom AMD APU/GPU), but if I didn't, I'm sorry. I heard the model noise is really a hard thing for the developers to fix, but let's keep hoping, because they're doing an awesome job!

**Comment by kazuya-555** [comment]
There`s also lights shinning thru walls in some levels, not sure if that`s simply Nvidia GPUs too, or is supposed to be like that.

And seems like the deadzone is larger than it needs to be, that makes using a game pad slightly more annoying than it already is. Yet since every game I tried isnt like this, then probably isnt Xenia doing that ? Would be nice if controller options where added for things like this though.

**Comment by Calinou** [comment]
> And seems like the deadzone is larger than it needs to be, with regular aiming slightly more than if using the LT. This makes using a gamepad slightly more annoying than it already is. Yet other games arnt like that, so probably isnt because of Xenia ?

This is an issue with the original game, not Xenia.

**Comment by madmodder123** [comment]
> And seems like the deadzone is larger than it needs to be, with regular aiming slightly more than if using the LT. This makes using a gamepad slightly more annoying than it already is. Yet other games arnt like that, so probably isnt because of Xenia ?

Are you playing with the latest Title Update? I think they might have fixed it, they messed around with some control settings

**Comment by GenesisAria** [comment]
I cannot get past the main menu when using a controller.  The XBLA profile popup never shows even if headless = false.  If i switch to winkey i get the popup and can proceed into the game.

**Comment by kazuya-555** [comment]
> Are you playing with the latest Title Update? I think they might have fixed it, they messed around with some control settings

I don`t think that updates load with Xenia right now ? I did try using the newer update, unless that`s corrected now.

**Comment by Sage009** [comment]
![20231019095322_1](https://github.com/xenia-project/game-compatibility/assets/32518671/54c282e3-90bc-4492-8b52-b02cf15f593a)
Got the game running in Xenia Canary on my Steam Deck, but the majority of the textures are messed up. This is supposed to be the very start of the first stage looking down. Where's the building I'm standing on?
Also Joanna herself and the table she's working on in the main menu are invisible.

**Comment by hooperre** [comment]
Appears to be using Vulkan. Try: `--gpu="d3d12"`

**Comment by Sage009** [comment]
> Appears to be using Vulkan. Try: `--gpu="d3d12"`

Switching to D3D12 solved the texture problem, but now I have a new problem: All menu text is garbled.
![20231023124916_1](https://github.com/xenia-project/game-compatibility/assets/32518671/286c64c2-6d07-4952-add1-d3001c9aa7c9)

**Comment by Sage009** [comment]
> > Switching to D3D12 solved the texture problem, but now I have a new problem: All menu text is garbled.
> 
> To avoid the broken menu just add the line "fullscreen=false" without quotes as start argument in game preferences.

Thanks! This solved the problem.

**Comment by Nelax88** [comment]
I solved the shader noise bug for Nvidia GPU users. The cause of the bug was the iso itself being unable to render the FogNearFar function on D3D12 adapters properly so changing the function integer from 1 to 0 fixed it. GPU Nvidia 2080 Super
![Fixed Result](https://github.com/xenia-project/game-compatibility/assets/122827832/f5d24fd9-befb-473b-809f-fbc9924fdc77)
Tested on Xenia Master Branch
Download fix: https://www.dropbox.com/scl/fi/8905i0ka7m5g0v1io4co7/Perfect-Dark-XBLA-NVIDIA-FIX.rar?rlkey=rh5k2yad5ya0d6pabbv9v1vvr&st=nmwfzkxt&dl=0
https://drive.google.com/file/d/1nU8NuDzASFgY30S_tf0ZiDG4-ycV7dFJ/view?usp=sharing

**Comment by TDOdev91** [comment]
![Fixed Result](https://github.com/xenia-project/game-compatibility/assets/172572966/6b5031a1-35df-4d0b-a94c-fdfa6f951466)
Tested on GPU Nvidia 2080 Super (WORKING!!!)

After many many hours of research and reviewing the Perfect Dark XBLA through assembly building software, I finally discovered the main cause of the bug for Nvidia GPUs and have tried to document as much as I can through my goal. Forewarn, I am not a reverse engineer specialists and have no knowledge or experience in it and this was done on a debug copy of the game which contains a few more bugs before final release. This was simply done through pure passion for the game, and dedication to countless hours of research. Nevertheless despite a few bugs on this copy, the game is now playable fully including combat simulator with the shader texture bug officially patched. I provided the fix version of the game here:
Google Drive: https://drive.google.com/file/d/1nU8NuDzASFgY30S_tf0ZiDG4-ycV7dFJ/view?usp=sharing
Dropbox: https://www.dropbox.com/scl/fi/8905i0ka7m5g0v1io4co7/Perfect-Dark-XBLA-NVIDIA-FIX.rar?rlkey=rh5k2yad5ya0d6pabbv9v1vvr&st=qzcn5ynd&dl=0

/Below is technical details and photos on how I discovered the bug, you can skip reading this if you only here for the fixed version\

The primary cause for the bug I discovered is due to the FogNearFar: register(c1) flag between address 000166F0 and 00016700 in the default.xex rendering improperly due to what I suspect is the result of the D3D12 adapter. I found this out after using HxD to manipulate the Float4 and Float3 functions and found out the "float4 FogCol: register(c4);\n" was the closes related to the shader causing the artifact on the dynamic models. Once I found this I went back further debugging any functions with the Fog flag. The shader bug simply went away by changing FogNearFar: register(c1) to FogNearFar: register(c0), and just like that it fixed!

![part 3](https://github.com/xenia-project/game-compatibility/assets/172572966/f65f5323-5f04-4656-92fe-513725ca0af8)
The culprit.

Again I'm not a careered reverse engineer, I merely did this because I love Perfect Dark and have been running on sheer passion to fix this, hopefully this makes everyones day. If anyone wishes to refer this to the Xenia development team, please refer them to this post and may someone with better skills than me be able to implement a patch because I don't think I'm up for this job xD

Game on, have a great year everyone!

**Comment by Graslu** [comment]
I can confirm the fix works! However now there's the issue of all lights going through geometry, so it's kinda unplayable either way. It's similar to how AMD renders it.

![image](https://github.com/xenia-project/game-compatibility/assets/34986384/1092bdd0-a3d4-4c8a-b62b-4ea6edc2b99a)

**Comment by BowserN64** [comment]
> ![Fixed Result](https://private-user-images.githubusercontent.com/172572966/339232749-6b5031a1-35df-4d0b-a94c-fdfa6f951466.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNjIzOTIsIm5iZiI6MTcxODI2MjA5MiwicGF0aCI6Ii8xNzI1NzI5NjYvMzM5MjMyNzQ5LTZiNTAzMWExLTM1ZGYtNGQwYi1hOTRjLWZkZmE2Zjk1MTQ2Ni5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxM1QwNzAxMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNzk5NjlmNDZmMzE2ZmMyMTQ2NmRkN2NjMjc3NWVmMWY3ZTk2YmNhZjM3YjEwMDcyMzFjZDgxNjAyYzhiMTM4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.tmVqpGcVh0j6_F6i3BjXCUmNnJQ0YEAM0i0MtyvNIqY) Tested on GPU Nvidia 2080 Super (WORKING!!!)
> 
> After many many hours of research and reviewing the Perfect Dark XBLA through assembly building software, I finally discovered the main cause of the bug for Nvidia GPUs and have tried to document as much as I can through my goal. Forewarn, I am not a reverse engineer specialists and have no knowledge or experience in it and this was done on a debug copy of the game which contains a few more bugs before final release. This was simply done through pure passion for the game, and dedication to countless hours of research. Nevertheless despite a few bugs on this copy, the game is now playable fully including combat simulator with the shader texture bug officially patched.
> 
> /Below is technical details and photos on how I discovered the bug\
> 
> The primary cause for the bug I discovered is due to the FogNearFar: register(c1) flag between address 000166F0 and 00016700 in the default.xex rendering improperly due to what I suspect is the result of the D3D12 adapter. I found this out after using HxD to manipulate the Float4 and Float3 functions and found out the "float4 FogCol: register(c4);\n" was the closes related to the shader causing the artifact on the dynamic models. Once I found this I went back further debugging any functions with the Fog flag. The shader bug simply went away by changing FogNearFar: register(c1) to FogNearFar: register(c0), and just like that it fixed!
> 
> ![part 3](https://private-user-images.githubusercontent.com/172572966/339232924-f65f5323-5f04-4656-92fe-513725ca0af8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNjIzOTIsIm5iZiI6MTcxODI2MjA5MiwicGF0aCI6Ii8xNzI1NzI5NjYvMzM5MjMyOTI0LWY2NWY1MzIzLTVmMDQtNDY1Ni05MmZlLTUxMzcyNWNhMGFmOC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxM1QwNzAxMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZjM2ZWFhNzMyMjU0MWE0ZWIzMTAwMDFiNjI5ZGRjNzQ3MTBiNWFkMGM0MzBhYTRjNDBkMDkxZDg3NDYzM2UwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.dBPJcP7en457LNc54SdYd6HMAr4higvagZcPZsBNP7s) The culprit.
> 
> Again I'm not a careered reverse engineer, I merely did this because I love Perfect Dark and have been running on sheer passion to fix this, hopefully this makes everyones day. If anyone wishes to refer this to the Xenia development team, please refer them to this post and may someone with better skills than me be able to implement a patch because I don't think I'm up for this job xD
> 
> Game on, have a great year everyone!

Wow, I...really don't know what to say. Fantastic work on solving this seemingly unsolvable, illusive issue which has plagued this game for years! ❤️❤️❤️🥳🎶🎉💥👊😺😻 Your fix INDEED works on my NVIDIA laptop! I will try it on my Desktop (NVIDIA RTX 4070 TI) when I get home! One thing I have noticed in my testing on DataDyne Defection is Graslu's observation about the lighting shining through everything, and also that shooting cameras only shows an explosion but the camera looks the same instead of being deformed as normal. I would assume this would be inherent through all other levels of the game. After all the work you described doing I feel bad to point out any issues at all, but these are MINOR concerns compared to jarring glitches like the noisy textures! I recently build a gaming PC and almost went AMD solely for playing this game with correct textures, so I am INDESCRIBABLY HAPPY you are working on fixing this issue for us NVIDIA users!!!! You, are a PRODIGY! And like super awesome in the highest sense of the word EVER! Super die-hard Perfect Dark fan here. God bless you! This is truly a momentous occasion in history for this game! Thank you so much again!!!! ❤️❤️❤️

**Comment by TDOdev91** [comment]
I've forwarded a majority of my findings to this repo here: https://github.com/xenia-canary/xenia-canary/pull/268
We'll be looking into creating a patch for the Canary version here very shortly and have this issue finally remediated for Nvidia users.

**Comment by TDOdev91** [comment]
Patches have been created and will be uploaded here very soon. Special major thanks to Adrian for creating and developing the patches to the game!

**Comment by BowserN64** [comment]
> Patches have been created and will be uploaded here very soon. Special major thanks to Adrian for creating and developing the patches to the game!

That's so awesome! You guys rock! 🎸🎶👊😎

**Comment by parabolee** [comment]
This fix is amazing! Thank you! Has anyone been able to get the game even running with a title update? If I install a title update I get an error with or without this patch.

![image](https://github.com/xenia-project/game-compatibility/assets/25558830/ed058197-c806-4ac6-8a01-da592b7098b8)

**Comment by parabolee** [comment]
OK so I just had the wrong title update for my version of the game. When or how can we make the Nvidia patch work with the latest title update? Thanks!

**Comment by parabolee** [comment]
> I can confirm the fix works! However now there's the issue of all lights going through geometry, so it's kinda unplayable either way. It's similar to how AMD renders it. Tested on a RTX 4090.
> 
> ![image](https://private-user-images.githubusercontent.com/34986384/339237208-1092bdd0-a3d4-4c8a-b62b-4ea6edc2b99a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2MDgzODYsIm5iZiI6MTcxODYwODA4NiwicGF0aCI6Ii8zNDk4NjM4NC8zMzkyMzcyMDgtMTA5MmJkZDAtYTNkNC00YzhhLWI2MmItNGVhNmVkYzJiOTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE3VDA3MDgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiZWI3NDFhNTQwYWQwYWExMDk3ZWE5NzBmMDVmMzQyNTc1N2M3YWQ1OGQ0M2ViNzE2NzhhNjBkZmU0NzI2MGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AfX0b4-56waM0oOcbFTC_GBBES5aicQ2WoELAVyjpAA)

Just an FYI, there is a "fix" for the light issue from a post earlier.

"Setting query_occlusion_fake_sample_count = 0 or -1 (https://github.com/xenia-project/game-compatibility/issues/255#issuecomment-1454886230) makes the lights not shine through the walls. Unfortunately, it does this by making them not shine even when you're looking straight at them, so it's not an actual fix."

**Comment by TDOdev91** [comment]
> OK so I just had the wrong title update for my version of the game. When or how can we make the Nvidia patch work with the latest title update? Thanks!

If you check the GItHub repo for Xenia Canary patches, a patch for the latest title update was just finished by Adrian and waiting to be merged to the main repository.

**Comment by parabolee** [comment]
Anyway to also get the mouse hook to work with TU3? Or we need the makers of it to patch support in I am guessing. Was going to see if that was a good way to reduce the dead zone issues. I am using Steam to it but it still doesn't feel anywhere near as nice or smooth as the PC port. I think I can get it really close though. Having to choose between the awesome updated graphics and the better feeling controls is rough though :)

**Comment by BrainyCabde** [comment]
> OK so I just had the wrong title update for my version of the game. When or how can we make the Nvidia patch work with the latest title update? Thanks!

Hey there parabolee, I see there's only 2 updates for Perfect Dark. V2 and V3. When I install either of them, it crashes the game with the same message you're getting there. Is there a specific way i'm supposed to install these updates to get the game working? Any help with this would be greatly appreciated! Thank you!

**Comment by madmodder123** [comment]
> > OK so I just had the wrong title update for my version of the game. When or how can we make the Nvidia patch work with the latest title update? Thanks!
> 
> Hey there parabolee, I see there's only 2 updates for Perfect Dark. V2 and V3. When I install either of them, it crashes the game with the same message you're getting there. Is there a specific way i'm supposed to install these updates to get the game working? Any help with this would be greatly appreciated! Thank you!

If you use an extractor (i forget which) you can extract all the files from the base game and the TU and put them in the same base game folder. you can then use an xex patch tool to patch the main xex file with the one in the TU.

That way you don't need to mess around hoping that title updates work properly

**Comment by BrainyCabde** [comment]
I got the correct Title Update it looks like, but my game name is different. It's 35C1CDD22DD0D4E54B858859C0052124FFFAD17958
Looks like i'll have to pass on using the updates. Thank you though!

**Comment by BrainyCabde** [comment]
Would I install both title updates, or just the latest one?

**Comment by parabolee** [comment]
Yeah the IR scanner and night vision have a few quirks, none that have hindered gameplay for me so far though.

**Comment by BrainyCabde** [comment]
Yeah, no hindered gameplay or nothing. You do have to know where that door is by playing the game without that bug though. So for new players, it would be a challenge. Just spam the open door button in the hallway and you'll find it though.

**Comment by Margen67** [comment]
Xenia Canary issue: https://github.com/xenia-canary/game-compatibility/issues/17
